packages/sls-env/src/index.ts

```diff
@@ -244,41 +246,40 @@ const environment = <H extends Handler<any, any, any>, C, D, P = HandlerPayload<
           })
-          // resolve dependencies
+          // create payload
           .then(({ logger, config, invocationId }) => {
+            logger.debug('about to create payload')
+            return Promise.resolve(payloadConstructor(event, context, { logger, invocationId })).then(payload => ({
+              config,
+              payload,
+              logger,
+              invocationId
+            }))
+          })
+          .then(({ logger, config, payload, invocationId }) => {
+            // we use trace here because dependencies could have sensitive values
+            logger.trace({ payload }, 'created payload')
+            return { logger, config, payload, invocationId }
+          })
+          // resolve dependencies
+          .then(({ logger, config, payload, invocationId }) => {
             logger.trace('about to resolve dependencies')
-            if (!applicationDependencies) {
+            if (!applicationDependencies || dependenciesConfig.cache === false) {
               logger.info('creating dependencies')
               if (typeof dependencies === 'function' && dependencies instanceof Function) {
-                applicationDependencies = dependencies({ config, logger, invocationId })
+                applicationDependencies = dependencies({ config, payload, logger, invocationId })
               } else {
                 logger.trace('about to resolve dependencies')
                 applicationDependencies = dependencies
               }
             }
 
-            return { logger, config, dependencies: applicationDependencies, invocationId }
+            return { logger, config, payload, dependencies: applicationDependencies, invocationId }
           })
-          .then(({ logger, config, dependencies, invocationId }) => {
+          .then(({ logger, config, payload, dependencies, invocationId }) => {
             // we use trace here because dependencies could have sensitive values
             logger.trace({ dependencies }, 'resolved dependencies')
-            return { logger, config, dependencies, invocationId }
-          })
-          // create payload
-          .then(({ logger, config, dependencies, invocationId }) => {
-            logger.debug('about to create payload')
-            return Promise.resolve(payloadConstructor(event, context, { logger, invocationId })).then(payload => ({
-              payload,
-              dependencies,
-              config,
-              logger,
-              invocationId
-            }))
-          })
-          .then(({ logger, config, dependencies, payload, invocationId }) => {
-            // we use trace here because dependencies could have sensitive values
-            logger.trace({ payload }, 'created payload')
-            return { logger, config, dependencies, payload, invocationId }
+            return { logger, config, payload, dependencies, invocationId }
           })
           // invoke application
```

Comment:
resurrect `create payload` and we would create another `thenable` and have inject `EventBasedDependency`

loop through dependencies and extract all `EventBasedDependency` and replace with `source.key` (where a source key is the key from the `event` )  based on `event`

```js

```